### PR TITLE
[RNMobile] Add disabled style to `Cell` component

### PR DESCRIPTION
**Pull request by fluiddot**
## What?

Add disabled style to `Cell` component to reflect when a cell is disabled.

## Why?

Disabling the `Cell` component didn't reflect any change in the UI, which is confusing to users.

## How?

Change the opacity of the value component (`Text` / `TextInput` components) when the `Cell` component is disabled.

Since this change might affect the different control components we use in the app, I checked the different components exported in the `@wordpress/components` package. The ones with ✅ are controls affected that have been verified:
- `TextControl` ✅
- `ColorControl` ✅
- `RangeControl - Slider` ✅
- `RangeControl - Stepper` ✅
- `ToggleControl` ✅ (This PR reverts https://github.com/WordPress/gutenberg/pull/50560 as now the disabled state is managed in the `Cell` component)
- `BottomSheetTextControl` ✅
- `BottomSheetSelectControl` ✅
- `CycleSelectControl` ✅
- `BaseControl` (cell not used)
- `TextareaControl` (cell not used)
- `SearchControl` (cell not used)
- `SelectControl` ✅
- `FooterMessageControl` (`disabled` not needed)
- `QueryControls` (composition of control components - `SelectControl` + `SelectControl` (`TreeSelect`) + `RangeControl`)
- `RadioControl` ✅
- `UnitControl` ✅
- `SegmentedControls` (cell not used)
- `SegmentedControl` (cell not used)

## Testing Instructions

In order to test the different control components, I created a test component for this purpose.

1. Apply the following patch to incorporate the test component:

<details><summary>Click here to display patch</summary>

```patch
diff --git forkSrcPrefix/packages/edit-post/src/components/layout/index.native.js forkDstPrefix/packages/edit-post/src/components/layout/index.native.js
index a9c38ee1ccb8d5ea579a46bde8070edc372e8567..e0d2ea1c969805500405bbeb8c75255ebc08793b 100644
--- forkSrcPrefix/packages/edit-post/src/components/layout/index.native.js
+++ forkDstPrefix/packages/edit-post/src/components/layout/index.native.js
@@ -33,6 +33,7 @@ import headerToolbarStyles from '../header/header-toolbar/style.scss';
 import Header from '../header';
 import VisualEditor from '../visual-editor';
 import { store as editPostStore } from '../../store';
+import TestControls from './test-controls';
 
 class Layout extends Component {
 	constructor() {
@@ -149,6 +150,7 @@ class Layout extends Component {
 				>
 					<AutosaveMonitor disableIntervalChecks />
 					<View style={ editorStyles }>
+						<TestControls />
 						{ isHtmlView ? this.renderHTML() : this.renderVisual() }
 						{ ! isHtmlView && Platform.OS === 'android' && (
 							<FloatingToolbar />
diff --git forkSrcPrefix/packages/edit-post/src/components/layout/test-controls.js forkDstPrefix/packages/edit-post/src/components/layout/test-controls.js
new file mode 100644
index 0000000000000000000000000000000000000000..7dacdb799c72f42449bcd0d004f4352a5f6a0187
--- /dev/null
+++ forkDstPrefix/packages/edit-post/src/components/layout/test-controls.js
@@ -0,0 +1,221 @@
+/**
+ * WordPress dependencies
+ */
+import {
+	PanelBody,
+	BottomSheet,
+	BottomSheetTextControl,
+	BottomSheetSelectControl,
+	FooterMessageControl,
+	RangeControl,
+	TextControl,
+	ColorControl,
+	ToggleControl,
+	CycleSelectControl,
+	SelectControl,
+	RadioControl,
+	UnitControl,
+	ColorSettings,
+} from '@wordpress/components';
+import { useState } from '@wordpress/element';
+import { useMultipleOriginColorsAndGradients } from '@wordpress/block-editor';
+
+/**
+ * External dependencies
+ */
+// eslint-disable-next-line import/no-extraneous-dependencies
+import { useNavigation } from '@react-navigation/native';
+
+const TestControlsContainer = () => {
+	const colorSettings = useMultipleOriginColorsAndGradients();
+
+	return (
+		<BottomSheet
+			isVisible={ true }
+			onClose={ () => {} }
+			hideHeader
+			hasNavigation
+		>
+			<BottomSheet.NavigationContainer animate main>
+				<BottomSheet.NavigationScreen name={ 'main' }>
+					<TestControls />
+				</BottomSheet.NavigationScreen>
+				<BottomSheet.NavigationScreen
+					name={ BottomSheet.SubSheet.screenName }
+				>
+					<BottomSheet.SubSheet.Slot />
+				</BottomSheet.NavigationScreen>
+				<BottomSheet.NavigationScreen name="Color">
+					<ColorSettings defaultSettings={ colorSettings } />
+				</BottomSheet.NavigationScreen>
+			</BottomSheet.NavigationContainer>
+		</BottomSheet>
+	);
+};
+
+const TestControls = () => {
+	const [ controlsDisabled, setControlsDisabled ] = useState( true );
+
+	const [ selectValue, setSelectValue ] = useState( 1 );
+	const [ rangeValue, setRangeValue ] = useState( 0 );
+	const [ unitValue, setUnitValue ] = useState( 'px' );
+	const [ textValue, setTextValue ] = useState( 'example' );
+	const [ colorValue, setColorValue ] = useState( '#000000' );
+	const [ toggleValue, setToggleValue ] = useState( false );
+
+	const navigation = useNavigation();
+
+	return (
+		<>
+			<PanelBody title="Panel only for Testing">
+				<ToggleControl
+					label="Disable the below controls"
+					onChange={ setControlsDisabled }
+					checked={ controlsDisabled }
+				/>
+			</PanelBody>
+			<PanelBody title="Select Controls">
+				<BottomSheetSelectControl
+					label="BottomSheetSelectControl"
+					options={ [
+						{ value: 1, label: 'Value 1' },
+						{ value: 2, label: 'Value 2' },
+						{ value: 3, label: 'Value 3' },
+					] }
+					onChange={ setSelectValue }
+					value={ selectValue }
+					disabled={ controlsDisabled }
+				/>
+				<CycleSelectControl
+					label="CycleSelectControl"
+					onChange={ setSelectValue }
+					options={ [
+						{ value: 1, name: 'Value 1' },
+						{ value: 2, name: 'Value 2' },
+						{ value: 3, name: 'Value 3' },
+					] }
+					value={ selectValue }
+					disabled={ controlsDisabled }
+				/>
+				<SelectControl
+					label="SelectControl"
+					value={ selectValue }
+					onChange={ setSelectValue }
+					options={ [
+						{ value: 1, label: 'Value 1' },
+						{ value: 2, label: 'Value 2' },
+						{ value: 3, label: 'Value 3' },
+					] }
+					hideCancelButton
+					disabled={ controlsDisabled }
+				/>
+			</PanelBody>
+			<PanelBody title="Radio Control">
+				<RadioControl
+					selected={ selectValue }
+					options={ [
+						{ value: 1, label: 'Value 1' },
+						{ value: 2, label: 'Value 2' },
+						{ value: 3, label: 'Value 3' },
+					] }
+					onChange={ setSelectValue }
+					disabled={ controlsDisabled }
+				/>
+			</PanelBody>
+			<PanelBody title="Range Controls">
+				<RangeControl
+					label="RangeControl - Stepper"
+					value={ rangeValue }
+					onChange={ setRangeValue }
+					min={ 0 }
+					max={ 10 }
+					type="stepper"
+					disabled={ controlsDisabled }
+				/>
+				<RangeControl
+					label="RangeControl - Slider"
+					value={ rangeValue }
+					onChange={ setRangeValue }
+					min={ 0 }
+					max={ 10 }
+					disabled={ controlsDisabled }
+				/>
+				<UnitControl
+					label="UnitControl"
+					min={ 0 }
+					max={ 10 }
+					value={ rangeValue }
+					onChange={ setRangeValue }
+					onUnitChange={ setUnitValue }
+					unit={ unitValue }
+					units={ [
+						{
+							value: 'px',
+							label: 'px',
+							step: 1,
+						},
+						{
+							value: 'em',
+							label: 'em',
+							step: 0.1,
+						},
+					] }
+					disabled={ controlsDisabled }
+				/>
+			</PanelBody>
+			<PanelBody title="Text Controls">
+				<TextControl
+					value={ textValue }
+					onChange={ setTextValue }
+					placeholder="placeholder"
+					label="TextControl"
+					disabled={ controlsDisabled }
+				/>
+				<TextControl
+					value=""
+					onChange={ setTextValue }
+					placeholder="placeholder"
+					label="TextControl"
+					disabled={ controlsDisabled }
+				/>
+				<BottomSheetTextControl
+					initialValue={ textValue }
+					onChange={ setTextValue }
+					label="BottomSheetTextControl"
+					disabled={ controlsDisabled }
+				/>
+				<BottomSheetTextControl
+					onChange={ setTextValue }
+					placeholder="placeholder"
+					label="BottomSheetTextControl"
+					disabled={ controlsDisabled }
+				/>
+			</PanelBody>
+			<PanelBody title="Other controls">
+				<ColorControl
+					label="ColorControl"
+					color={ colorValue }
+					onPress={ () => {
+						navigation.navigate( 'Color', {
+							onColorChange: setColorValue,
+							colorValue,
+							onColorCleared: () => setColorValue( undefined ),
+							label: 'ColorControl',
+						} );
+					} }
+					withColorIndicator
+					disabled={ controlsDisabled }
+				/>
+				<ToggleControl
+					label="ToggleControl"
+					onChange={ setToggleValue }
+					checked={ toggleValue }
+					disabled={ controlsDisabled }
+				/>
+				<FooterMessageControl label="FooterMessageControl" />
+			</PanelBody>
+		</>
+	);
+};
+
+export default TestControlsContainer;

```

</details> 

2. Open a post/page in the app.
3. Observe that a bottom sheet is automatically open.
4. Observe that all displayed controls are disabled.
5. Enable the controls using the toggle located at the top of the sheet.
6. Observe that all displayed controls are enabled.

### Testing Instructions for Keyboard

N/A

## Screenshots or screencast 

<details><summary><strong>Click here to show Android screenshots</strong></summary>

## Light mode
| Cell active | Cell disabled |
|--------|--------|
|   |  | 
|  |  | 
|  |  |

## Dark mode
| Cell active | Cell disabled |
|--------|--------|
|   |  | 
|  |  | 
|  |  |

</details> 

<details><summary><strong>Click here to show iOS screenshots</strong></summary>

## Light mode
| Cell active | Cell disabled |
|--------|--------|
|   |  | 
|  |  | 
|  |  |

## Dark mode
| Cell active | Cell disabled |
|--------|--------|
|   |  | 
|  |  | 
|  |  |

</details>

### CODE REVIEW

File: packages/components/src/mobile/bottom-sheet/cell.native.js
Comment (fluiddot) [inline packages/components/src/mobile/bottom-sheet/cell.native.js:266]: The text input is also disabled.

File: packages/components/src/mobile/bottom-sheet/range-cell.native.js
Comment (fluiddot) [inline packages/components/src/mobile/bottom-sheet/range-cell.native.js:228]: On iOS the default disabled style of the Slider already makes it transparent. So, in order to avoid making it too transparent with the Cell's disabled style, we only disable it on Android.

File: packages/components/src/mobile/bottom-sheet/switch-cell.native.js
Comment (fluiddot) [inline packages/components/src/mobile/bottom-sheet/switch-cell.native.js:66]: In the case of the `Switch` component, we don't want to make it transparent as it already provides a disabled style via the `disabled` prop. Hence, we pass an empty style to override it.

File: packages/components/src/mobile/bottom-sheet/cell.native.js
Comment (fluiddot) [inline packages/components/src/mobile/bottom-sheet/cell.native.js:439]: Although the parent component (`TouchableRipple`) is disabled, I noticed that some controls render input elements that also need to be disabled to prevent interaction. Instead of disabling them one by one, I simply disabled all touch events for the children when the cell is disabled.

**Comment by fluiddot** [comment]
~~Heads up that I noticed that the placeholder color for `BottomSheetTextControl` component wasn't being affected by the disabled styles. I pushed a fix for this in https://github.com/WordPress/gutenberg/pull/50665/commits/d51768d7a7ec13a7e024efa9896807444b447702.~~

**UPDATE:** I've applied another fix as shared in https://github.com/WordPress/gutenberg/pull/50665#issuecomment-1551010124.

**Comment by fluiddot** [comment]
> Heads up that I noticed that the placeholder color for `BottomSheetTextControl` component wasn't being affected by the disabled styles. I pushed a fix for this in [d51768d](https://github.com/WordPress/gutenberg/commit/d51768d7a7ec13a7e024efa9896807444b447702).

I noticed some issues when testing this change, so I've edited the changes and pushed another fix in https://github.com/WordPress/gutenberg/pull/50665/commits/4d9fd1551b58e222bb2cd4c5c2dd7c481d80b735.

Heads up that I've also updated the screenshots in the PR's description to reflect the last changes.

File: packages/components/src/mobile/bottom-sheet/cell.native.js
Comment (fluiddot) [inline packages/components/src/mobile/bottom-sheet/cell.native.js:237]: I finally calculated the placeholder text color to handle the different cases.

File: packages/components/src/mobile/bottom-sheet/styles.native.scss
Comment (fluiddot) [inline packages/components/src/mobile/bottom-sheet/styles.native.scss:287]: I noticed this color is already defined globally:
https://github.com/WordPress/gutenberg/blob/6f2d112399b007ccf33e913183e36446d841c9f7/packages/base-styles/_colors.native.scss#L41

**Comment by fluiddot** [comment]
@osullivanchris I'd appreciate your design input on these changes (I added several screenshots in the PR's description). I haven't generated an installable build, but let me know if you'd like to test them with a build and I'll create one.
Thanks for the help 🙇 !

File: packages/components/src/mobile/bottom-sheet/cell.native.js
Comment (derekblank) [reply packages/components/src/mobile/bottom-sheet/cell.native.js:237]: This is great. I noticed this when I started my review and then noted you updated this code. :) The raw hex values stood out to me even when working on the [Clear Button issue](https://github.com/WordPress/gutenberg/pull/37402) from a while ago -- thanks for addressing them here!

**Review by derekblank — APPROVED** [review]
This looks great, thanks for tackling this! I was able to test disabled inputs on both platforms successfully, including dark mode on both. Additionally, thank you for adding the patch to the testing instructions, which made this very easy to test. LGTM. 🚀

**Comment by fluiddot** [comment]
Thank @derekblank for reviewing the PR 🙇 !

> [...] Additionally, thank you for adding the patch to the testing instructions, which made this very easy to test. LGTM. 🚀

I also found having the test component particularly useful when testing. In fact, I'm thinking of incorporating it into the code base for testing purposes. We could even use Storybook for this, although it might require extra work to support RN.

**Comment by osullivanchris** [comment]
@fluiddot I might take a closer look at this one. If you want to proceed, and if I have an improvement in mind I can let you know.

A couple of things come to mind:
- the disabled state for some items have a similar but slightly different color. As its using transparency. In [this screenshot](https://user-images.githubusercontent.com/14905380/238889058-92b3210d-f1ef-4634-99e2-7023309ecdca.jpg) for example. There's grey-ish and blue-ish grey text. Maybe a hard coded disabled color would be better. Or maybe its a question higher up the chain of why our labels are sometimes a blue-ish grey and sometimes a flat grey color.
- I think when a ColorControl is disabled it should be a special case and not make the colour transparent. It can make it unclear what the colour in use is.
- I'm trying to look at other examples to see if the field name as well as the field value should have a disabled state. Right now the field name on the left always remains in an enabled appearance. Maybe fine. Just considering.
- Should we show a toast message or similar when a user taps a disabled field, to add clarity?

**Comment by fluiddot** [comment]
> @fluiddot I might take a closer look at this one. If you want to proceed, and if I have an improvement in mind I can let you know.

Great, thanks @osullivanchris for the feedback 🙇 ! Let me know if there's any improvement that we should make before merging the PR. These changes are blocking https://github.com/wordpress-mobile/gutenberg-mobile/pull/5782, but it can wait until we provide a good solution here. Note that the disabled state is not yet used, it will be first used by the changes introduced in https://github.com/wordpress-mobile/gutenberg-mobile/pull/5782.

> * the disabled state for some items have a similar but slightly different color. As its using transparency. In [this screenshot](https://user-images.githubusercontent.com/14905380/238889058-92b3210d-f1ef-4634-99e2-7023309ecdca.jpg) for example. There's grey-ish and blue-ish grey text. Maybe a hard coded disabled color would be better. Or maybe its a question higher up the chain of why our labels are sometimes a blue-ish grey and sometimes a flat grey color.

The blue-ish is used as the placeholder color (i.e. when the text field is empty) while the grey-ish is for regular text. I understand the color difference is to differentiate when a text field is empty. Should we use the same color when the component is disabled?

> * I think when a ColorControl is disabled it should be a special case and not make the colour transparent. It can make it unclear what the colour in use is.

Good point, I can handle that case independently to prevent making the color transparent. However, I'm wondering if the disabled state would be noticeable just by adding transparency to the right arrow. WDYT?

> * I'm trying to look at other examples to see if the field name as well as the field value should have a disabled state. Right now the field name on the left always remains in an enabled appearance. Maybe fine. Just considering.

We could also apply a disabled style to the field name, I was tempted to make it also transparent but I felt it was too much transparency. Do you have any idea what we could do in this case? This could help the case I mentioned above about the `ColorControl`.

> * Should we show a toast message or similar when a user taps a disabled field, to add clarity?

Good idea. This could be helpful to clarify why the field is disabled. I'll open a GitHub issue as a follow-up after we merge the PR.

**Comment by fluiddot** [comment]
@osullivanchris As we discussed internally, I'm going to merge this as-is to unblock the changes from https://github.com/wordpress-mobile/gutenberg-mobile/pull/5782. I'll incorporate the design changes we discussed in a separate PR.

**Comment by fluiddot** [comment]
> @osullivanchris As we discussed internally, I'm going to merge this as-is to unblock the changes from [wordpress-mobile/gutenberg-mobile#5782](https://github.com/wordpress-mobile/gutenberg-mobile/pull/5782). I'll incorporate the design changes we discussed in a separate PR.

[Here is the PR](https://github.com/WordPress/gutenberg/pull/50907) with the changes.

**Pull request merged**
